contrib/zenith/pagestore_prefetch.c

```diff
@@ -0,0 +1,401 @@
+static bool prefetch_cancel;
+
+size_t n_prefetch_requests;
+size_t n_prefetch_hits;
+size_t max_merged_prefetch_requests;
+
+static void
+zenith_prefetch_shmem_startup(void)
+{
+	bool found;
+
+	if (prev_shmem_startup_hook) {
+		prev_shmem_startup_hook();
+    }
+
+	LWLockAcquire(AddinShmemInitLock, LW_EXCLUSIVE);
+	prefetch_buffer = ShmemInitStruct("zenithr_prefetch",
+									  sizeof(PrefetchControl) +
+									  (BLCKSZ+sizeof(PrefetchEntry*))*prefetch_buffer_size,
+									  &found);
+	prefetch_control = (PrefetchControl*)(prefetch_buffer + BLCKSZ*(size_t)prefetch_buffer_size);
+	if (!found)
+	{
+		static HASHCTL info;
+		prefetch_lock = (LWLockId)GetNamedLWLockTranche("zenith_prefetch");
+		info.keysize = sizeof(BufferTag);
+		info.entrysize = sizeof(PrefetchEntry);
+		prefetch_hash = ShmemInitHash("zenith_prefetch",
+									  prefetch_buffer_size, prefetch_buffer_size,
+									  &info,
+									  HASH_ELEM | HASH_BLOBS);
+		prefetch_control->curr = 0;
+		prefetch_control->go_latch = NULL;
+		memset(prefetch_control->entries, 0, prefetch_buffer_size*sizeof(PrefetchEntry*));
+	}
+	LWLockRelease(AddinShmemInitLock);
+}
+
+void
+zenith_prefetch_init(void)
+{
+	/*
+	 * In order to create our shared memory area, we have to be loaded via
+	 * shared_preload_libraries.
+	 */
+	if (!process_shared_preload_libraries_in_progress)
+		elog(ERROR, "Zenith module should be loaed via shared_preload_libraries");
+
+	DefineCustomIntVariable("zenith.prefetch_buffer_size",
+                            "Size of zenith prefetch buffer",
+							NULL,
+							&prefetch_buffer_size,
+							1024, /* 8Mb */
+							0,
+							INT_MAX,
+							PGC_POSTMASTER,
+							0,
+							NULL,
+							NULL,
+							NULL);
+
+	if (prefetch_buffer_size == 0)
+		return;
+
+	RequestAddinShmemSpace((size_t)prefetch_buffer_size*(BLCKSZ+sizeof(PrefetchEntry*))
+						   + sizeof(PrefetchControl)
+						   + hash_estimate_size(prefetch_buffer_size, sizeof(PrefetchEntry)));
+	RequestNamedLWLockTranche("zenith_prefetch", 1);
+
+	prev_shmem_startup_hook = shmem_startup_hook;
+	shmem_startup_hook = zenith_prefetch_shmem_startup;
+}
+
+
+bool zenith_find_prefetched_buffer(SMgrRelation reln, ForkNumber forknum, BlockNumber blocknum, XLogRecPtr lsn, char* buffer)
+{
+	PrefetchEntry* entry;
+	BufferTag tag;
+
+	if (prefetch_buffer_size == 0)
+		return false;
+
+	tag.rnode = reln->smgr_rnode.node;
+	tag.forkNum = forknum;
+	tag.blockNum = blocknum;
+
+	while (true)
+	{
+		LWLockAcquire(prefetch_lock, LW_EXCLUSIVE);
+		entry = hash_search(prefetch_hash, &tag, HASH_FIND, NULL);
+		if (entry != NULL) {
+			if (entry->state == QUEUED || entry->lsn < lsn)
+			{
+				/* We have not sent this prefetch request and page is already requested.
+				 * So just concel this prefetch request.
+				 */
+				prefetch_control->entries[entry->index] = NULL;
+				hash_search(prefetch_hash, &tag, HASH_REMOVE, NULL);
+				if (entry->wait_latch)
+					SetLatch(entry->wait_latch);
```

Comment:
Technically, this is a `use-after-free` type of error. The correct sequence is used elsewhere in this patch; I'd extract it into a fuction.